resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/CharsetManager.java

```diff
@@ -421,46 +440,74 @@ public void projectPreferencesChanged(IProject project) {
-			if (newCharset != null)
-				resourcesPreferences.put(ResourcesPlugin.PREF_ENCODING, newCharset);
-			else
-				resourcesPreferences.remove(ResourcesPlugin.PREF_ENCODING);
-			try {
-				resourcesPreferences.flush();
-			} catch (BackingStoreException e) {
-				IProject project = workspace.getRoot().getProject(resourcePath.segment(0));
-				String message = Messages.resources_savingEncoding;
-				throw new ResourceException(IResourceStatus.FAILED_SETTING_CHARSET, project.getFullPath(), message, e);
-			}
-			return;
+	public void setCharsetFor(IPath resourcePath, String newCharset) throws CoreException, ResourceException {
+		if (isResourceRoot(resourcePath))
+			setCharsetForRoot(resourcePath, newCharset);
+		else
+			setCharsetForResource(resourcePath, newCharset, getResourceFromPath(resourcePath));
+	}
+
+	private void setCharsetForRoot(IPath resourcePath, String newCharset) throws ResourceException {
+		IEclipsePreferences resourcesPreferences = InstanceScope.INSTANCE.getNode(ResourcesPlugin.PI_RESOURCES);
+
+		if (newCharset != null)
+			resourcesPreferences.put(ResourcesPlugin.PREF_ENCODING, newCharset);
+		else
+			resourcesPreferences.remove(ResourcesPlugin.PREF_ENCODING);
+
+		try {
+			resourcesPreferences.flush();
+		} catch (BackingStoreException e) {
+			setCharsetForHasFailed(resourcePath, e);
 		}
-		// for all other cases, we set a property in the corresponding project
-		IResource resource = workspace.getRoot().findMember(resourcePath);
-		if (resource != null) {
-			try {
-				// disable the listener so we don't react to changes made by ourselves
-				Preferences encodingSettings = getPreferences(resource.getProject(), true, resource.isDerived(IResource.CHECK_ANCESTORS));
-				if (newCharset == null || newCharset.trim().length() == 0)
-					encodingSettings.remove(getKeyFor(resourcePath));
-				else
-					encodingSettings.put(getKeyFor(resourcePath), newCharset);
-				flushPreferences(encodingSettings, true);
-				if (resource instanceof IProject) {
-					IProject project = (IProject) resource;
-					ValidateProjectEncoding.scheduleProjectValidation(workspace, project);
-				}
-			} catch (BackingStoreException e) {
-				IProject project = workspace.getRoot().getProject(resourcePath.segment(0));
-				String message = Messages.resources_savingEncoding;
-				throw new ResourceException(IResourceStatus.FAILED_SETTING_CHARSET, project.getFullPath(), message, e);
+	}
+
+	private void setCharsetForResource(IPath resourcePath, String newCharset, IResource resource)
+			throws ResourceException {
+		try {
+			setResourceEncodingSettings(resourcePath, newCharset, resource);
+			if (resource instanceof IProject) {
+				IProject project = (IProject) resource;
+				ValidateProjectEncoding.scheduleProjectValidation(workspace, project);
 			}
+		} catch (BackingStoreException e) {
+			setCharsetForHasFailed(resourcePath, e);
 		}
 	}
 
+	private IResource getResourceFromPath(IPath resourcePath) throws ResourceException {
```

Comment:
As this method only has 1 consumer and is trivial to understand (only delegating to `workspace.getRoot().findMember()`, it would IMO be easier to inline and remove it.

Reply:
In a newer version of my code, the function is used like this:

```java	
public void setCharsetFor(IPath resourcePath, String newCharset) throws CoreException, ResourceException {
		if (IPath.ROOT.equals(resourcePath)) {
			setCharsetForRoot(resourcePath, newCharset);
		} else if (getResourceFromPath(resourcePath) instanceof Project project) {
			setCharsetForResource(resourcePath, newCharset, project);
		}
}```

I would argue that inlining this behaviour is not readily possible in this `if (xyz instanceof Project project)` without creating a very long line of code. Am I missing something?

Reply:
Depends on what you call very long. To me, as long as I can read it on my screen without scrolling and the line is constistent (not doing too much as once), it's fine.